### ISSUE TITLE
Results are not repeatable

### ISSUE THREAD

**Issue by lamyergeier** [maxim-saplin/NetCoreStorageSpeedTest#10]
Following test are conducted on same external harddrive on the same laptop from same port. We can observe that the test shows widely varying values for Sequential and random read! 

# Test 1
1/5 Sequential write [4MB] block        Avg: 45.46MB/s   Min÷Max: 21.86 ÷ 87.74, Time: 0m24s          
2/5 Sequential read [4MB] block         Avg: 426.19MB/s  Min÷Max: 76.47 ÷ 2570.87, Time: 0m03s        
3/5 Random write [4KB] block            Avg: 1.01MB/s    Min÷Max: 0.18 ÷ 8.83, Time: 0m07s            
4/5 Random read [4KB] block             Avg: 115.80MB/s  Min÷Max: 0.12 ÷ 1584.69, Time: 0m07s         
5/5 Memory copy [4MB] block             Avg: 3772.37MB/s         Min÷Max: 982.27 ÷ 10246.98, Time: 0m04s

# Test 2

1/5 Sequential write [4MB] block        Avg: 44.06MB/s   Min÷Max: 18.34 ÷ 79.22, Time: 0m25s          
2/5 Sequential read [4MB] block         Avg: 976.09MB/s  Min÷Max: 115.66 ÷ 3086.32, Time: 0m02s       
3/5 Random write [4KB] block            Avg: 1.05MB/s    Min÷Max: 0.18 ÷ 8.71, Time: 0m07s            
4/5 Random read [4KB] block             Avg: 3.71MB/s    Min÷Max: 0.11 ÷ 262.38, Time: 0m07s          
5/5 Memory copy [4MB] block             Avg: 3452.52MB/s         Min÷Max: 1131.59 ÷ 10590.42, Time: 0m04s

# Test 3

  1/5 Sequential write [4MB] block Avg: 44.97MB/s   Min÷Max: 20.61 ÷ 106.92, Time: 0m25s
  2/5 Sequential read [4MB] block  Avg: 527.74MB/s  Min÷Max: 81.72 ÷ 3118.75, Time: 0m03s
  3/5 Random write [4KB] block     Avg: 1.08MB/s    Min÷Max: 0.19 ÷ 8.68, Time: 0m07s
  4/5 Random read [4KB] block      Avg: 8.68MB/s    Min÷Max: 0.13 ÷ 293.81, Time: 0m07s
  5/5 Memory copy [4MB] block      Avg: 2214.92MB/s Min÷Max: 622.90 ÷ 8879.18, Time: 0m04s

**Comment by maxim-saplin** [comment]
Indeed, write tests are stable while read tests are not. Most likely that's related to cahced reads and OS ignoring apps request not to use file cache:
 
```
                if (RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform.Linux))
                    Syscall.posix_fadvise((int)SafeFileHandle.DangerousGetHandle(), 0, 0, PosixFadviseAdvice.POSIX_FADV_DONTNEED);
```

 There's no easy fix for that other than experimenting with various hacks and trying it out on your config to see which hack plays out and disables cached reads.

**Comment by lamyergeier** [comment]
If cache is the issue, am I correct in assuming that the first measurements are correct and subsequent measurements not for each device. Also, can cache affect the calculations of the first measurement itself?

21 May 2022 20:24:24 Maxim Saplin ***@***.***>:

**Comment by maxim-saplin** [comment]
That is not correct. Please notice that all read tests follow write ones. The write tests essentialy create the test file and thus let OS cache this file. Read tests do their measurements against file created by the very first Sequential Write test.

**Comment by lamyergeier** [comment]
Will it be better if the software creates a new test file after each write or read?

22 May 2022 11:24:30 Maxim Saplin ***@***.***>:

**Comment by maxim-saplin** [comment]
Software always creates a new file when starting a test session and deletes it afterwards. You can't workaround file caching with any sequences of creation of file.  The file is writen twice and read twice.

**Comment by lamyergeier** [comment]
Actually, I meant within a test session for each write and each read, the software could create a new test file.

22 May 2022 11:45:29 Maxim Saplin ***@***.***>:

**Comment by maxim-saplin** [comment]
And I tried to explain it doesn't matter how many files you create, as long as read tests do read what has just been written

**Comment by maxim-saplin** [comment]
Anyways, the best course of action is finding out how to make the OS ignore file cache and set up test file the proper way. There might be some distro specific API call.

For the moment I don't have capacity to follow-up with reproduction and investigation. Keeping the issue open
